### ISSUE TITLE
Using flag.conflicts causes issues with set -u

### ISSUE THREAD

**Issue by thendricks0** [DannyBen/bashly#559]
### Bashly Version

Latest Docker Image

### Description

Using flag.conflicts causes issues with set -u.
If I use strict mode or rather set -u and then try to use conflicting variables I get an unbound variable error since $key is unbound.

### Contents of bashly.yml

```yaml
flags:
  - long: --verbose
    short: -v
    help: Show verbose output
    conflicts: [--quiet]
  - long: --debug
    short: -d
    help: Show debug output
    conflicts: [--quiet]
  - long: --quiet
    short: -q
    help: Suppress output, this implies --force so all prompts are confirmed
    conflicts: [--verbose, --debug]
```

### Reproduction Steps

Use the provided bashly config + settings.yml below, generate a script and then execute the script with -qd or -qv or -qdv.

```yaml
# settings.yml

strict: true
# or use
strict: set -o nounset
```

### Actual Behavior

parse_requirements will be generated like this:
```bash
parse_requirements() {
  # :command.fixed_flags_filter
  while [[ $# -gt 0 ]]; do
    case "${1:-}" in
      --version)
        version_command
        exit
        ;;

      --help | -h)
        long_usage=yes
        awps_usage
        exit
        ;;

      # :flag.case
      --verbose | -v)
        # :flag.conflicts
        if [[ -n "${args['--quiet']:-}" ]]; then
          printf "conflicting options: %s cannot be used with %s\n" "$key" "--quiet" >&2
          exit 1
        fi

        # :flag.case_no_arg
        args['--verbose']=1
        shift
        ;;

      # :flag.case
      --debug | -d)
        # :flag.conflicts
        if [[ -n "${args['--quiet']:-}" ]]; then
          printf "conflicting options: %s cannot be used with %s\n" "$key" "--quiet" >&2
          exit 1
        fi

        # :flag.case_no_arg
        args['--debug']=1
        shift
        ;;

      # :flag.case
      --quiet | -q)
        # :flag.conflicts
        for conflict in --verbose --debug; do
          if [[ -n "${args[$conflict]:-}" ]]; then
            printf "conflicting options: %s cannot be used with %s\n" "$key" "$conflict" >&2
            exit 1
          fi
        done

        # :flag.case_no_arg
        args['--quiet']=1
        shift
        ;;
...
```

### Expected Behavior

But it should be generated like this:
```bash
parse_requirements() {
  # :command.fixed_flags_filter
  while [[ $# -gt 0 ]]; do
    key="$1"   # <----------- this!
    case "${key:-}" in
      --version)
        version_command
        exit
        ;;

      --help | -h)
        long_usage=yes
        awps_usage
        exit
        ;;

      # :flag.case
      --verbose | -v)
        # :flag.conflicts
        if [[ -n "${args['--quiet']:-}" ]]; then
          printf "conflicting options: %s cannot be used with %s\n" "$key" "--quiet" >&2
          exit 1
        fi

        # :flag.case_no_arg
        args['--verbose']=1
        shift
        ;;

      # :flag.case
      --debug | -d)
        # :flag.conflicts
        if [[ -n "${args['--quiet']:-}" ]]; then
          printf "conflicting options: %s cannot be used with %s\n" "$key" "--quiet" >&2
          exit 1
        fi

        # :flag.case_no_arg
        args['--debug']=1
        shift
        ;;

      # :flag.case
      --quiet | -q)
        # :flag.conflicts
        for conflict in --verbose --debug; do
          if [[ -n "${args[$conflict]:-}" ]]; then
            printf "conflicting options: %s cannot be used with %s\n" "$key" "$conflict" >&2
            exit 1
          fi
        done

        # :flag.case_no_arg
        args['--quiet']=1
        shift
        ;;
...
```

**Comment by DannyBen** [comment]
I can't reproduce the problem. Using the provided bashly.yml and `strict: true` in the settings produces a `cli` that outputs the expected result:

```shell
$ ./cli -qd
conflicting options: -d cannot be used with --quiet
```

and shellcheck does not have any complaint:

Are you on bashly 1.2.2?

The only place the generated script has the variable `$key` in it is here:

```bash
  # :command.parse_requirements_while
  while [[ $# -gt 0 ]]; do
    key="$1"
    case "$key" in
      # :flag.case
```

and I don't see how it can be unbound.

**Comment by thendricks0** [comment]
I use the latest bashly image as an alias.
I just created a new project using bashly init, pasted in the flags of the ticket and set the strict option in settings.yml to true.
Then I immediately get the unbound variable error:
```
$ ./cli -qdv download
./cli: line 299: key: unbound variable
```

This is how the parse_requirements function is generated:

```
# :command.parse_requirements
parse_requirements() {
  # :command.fixed_flags_filter
  while [[ $# -gt 0 ]]; do
    case "${1:-}" in
      --version)
        version_command
        exit
        ;;

      --help | -h)
        long_usage=yes
        cli_usage
        exit
        ;;

      # :flag.case
      --verbose | -v)
        # :flag.conflicts
        if [[ -n "${args['--quiet']:-}" ]]; then
          printf "conflicting options: %s cannot be used with %s\n" "$key" "--quiet" >&2
          exit 1
        fi

        # :flag.case_no_arg
        args['--verbose']=1
        shift
        ;;

      # :flag.case
      --debug | -d)
        # :flag.conflicts
        if [[ -n "${args['--quiet']:-}" ]]; then
          printf "conflicting options: %s cannot be used with %s\n" "$key" "--quiet" >&2
          exit 1
        fi

        # :flag.case_no_arg
        args['--debug']=1
        shift
        ;;

      # :flag.case
      --quiet | -q)
        # :flag.conflicts
        for conflict in --verbose --debug; do
          if [[ -n "${args[$conflict]:-}" ]]; then
            printf "conflicting options: %s cannot be used with %s\n" "$key" "$conflict" >&2
            exit 1
          fi
        done

        # :flag.case_no_arg
        args['--quiet']=1
        shift
        ;;

      *)
        break
        ;;

    esac
  done
```

`$key` is definitely unbound here.

The docker image uses bashly 1.2.2, I checked and the header of my generated script also says so

**Comment by thendricks0** [comment]
Your conflicts example only works as long as there are no commands set in the bashly.yml

As soon as I add commands the script is generated with the unbound variable:
```yaml
name: download
help: Sample application to demonstrate the use of conflicting flags
version: 0.1.0

flags:
- long: --cache
  help: Enable cache
  # Running --cache with --no-cache is not permitted
  conflicts: [--no-cache]
- long: --no-cache
  help: Disable cache
  # Running --no-cache with --cache or with --fast is not permitted
  conflicts: [--cache, --fast]
- long: --fast
  help: Run faster
  # Make sure to add the conflicting flags in both flags
  conflicts: [--no-cache]

commands:
- name: test
  alias: t
  help: Example to show parse_requirements is bugged
```

**Comment by thendricks0** [comment]
Thanks!

**Comment by DannyBen** [comment]
In fact, the problem is deeper and more specific.
This error occurs only when there is `conflicts` in the "fixed args" - meaning, when flags are defined alongside commands.

For example, this works:

```yaml
commands:
- name: test
  alias: t
  help: Example to show parse_requirements is bugged

  flags:
  - long: --fast
    short: -f
    conflicts: [--slow]
  - long: --slow
    short: -s
    conflicts: [--fast]
```

The fixed args `while` uses `$1` directly without setting `$key`.
Hopefully will be fixed in the next hour or so, it would help if you can test and verify once it is.

**Comment by thendricks0** [comment]
Sure! I'm currently re-writing a script using bashly and so far it has been a good experience and I actually need this to work, so I'll test as soon as you can provide a fix for it.

A little offtopic here, but I wondered if you know where I can find more libs for bashly? Using the github search doesn't provide very good results since "bashly" is fuzzy matched against "bash" so I get way too many results, google doesn't provide any good results. I thought since your project had 2k stars there would be more libs around, since I'm really missing some good logging library.
I wrote my own but I'm not really satisfied with it tbh.

**Comment by DannyBen** [comment]
Happy to hear it, and if something does not work as you expect, I am happy to help in your migration (issues / discussions are both fine).

As for libs for bashly - I don't know if there are any, but as you probably know by now - these are just functions that you place in `libs` - so you can search for general functions, or ask one of the AI chats to write you that function, and there you have it.

**Comment by DannyBen** [comment]
This should be fixed now in #560, merged to master, and deployed to the docker edge tag.

You can alias like this:

```shell
$ docker pull dannyben/bashly:edge
$ alias bashly-edge='docker run --rm -it --user $(id -u):$(id -g) --volume "$PWD:/app" dannyben/bashly:edge'
```

**Comment by thendricks0** [comment]
Works!
```
$ ./cli -qv
conflicting options: -q cannot be used with --verbose
$ ./cli -qd
conflicting options: -q cannot be used with --debug
```

**Comment by DannyBen** [comment]
Excellent. I want to fix #561 and then I will release (probably later today).

Thanks for the report.

**Comment by DannyBen** [comment]
Version 1.2.3 is live with this fix.
Thanks again for reporting.
